Tep 0114 required cancellation (#881)
tektoncd/community: teps/0114-custom-tasks-beta.md

last-updated: '2022-11-14'
authors:
- '@jerop'

> 

Reviewer: Please feel free to add yourself as a author ✍️

Author: Thanks @pritidesai~ It totally slipped from my mind.

When [TEP-0096: Pipelines V1 API][tep-0096] is implemented to add V1 API, `Custom Tasks` will be gated behind feature
gate `enable-api-fields` being set to `"beta"` - this is out of scope for this TEP  (in scope for TEP-0096).

##### Cancellation

> Implementing cancellation is now `required` instead of `optional`. If Custom Task Implementor does not implement `Cancellation`, pipeline-level timeout and PipelineRun cancellation will not work properly.

Reviewer 2: A bit of a wording suggestion: 

```suggestion
Implementing cancellation is now `required` instead of `optional`. If the Custom Task implementor does not support cancellation via `.spec.status`, `Pipeline` timeouts and `PipelineRun` cancellation will not work properly.
```

Author: Thanks @abayer. 

I reviewed TEP-0002 and it didn't explicitly tell users that how do we set the `spec.Status` and `spec.StatusMessage` and how do we expect custom task implementors signal back. I'll need to update this TEP to make that explicit.

When [TEP-0096: Pipelines V1 API][tep-0096] is implemented to add V1 API, `Custom Tasks` will be gated behind feature
gate `enable-api-fields` being set to `"beta"` - this is out of scope for this TEP  (in scope for TEP-0096).

##### Cancellation

> Implementing cancellation is now `required` instead of `optional`. If the Custom Task implementor does not support cancellation via `.spec.status`, `Pipeline` timeouts and `PipelineRun` cancellation will not work properly.

Reviewer: What does `properly` allude to here?

@ScrapCodes please share your experience here if you have any

Reviewer 3: To ensure that cancellation is required, are we planning to have validations now or in future? if no, is this just a documentation change?

Also, do we support pipeline level Timeout for custom task? I remember it was removed? 

What are the downside of keeping as: "It is upto the custom task controller to support it." i.e. optional, e.g. not all custom task may want to support it e.g. CEL. Those custom task who do not wish to support they are responsible for not supporting timeout as well as of today.

Today `Tekton controller` does not depend on custom task's support of cancel, is it going to change? 

>`Pipeline` timeouts and `PipelineRun` cancellation will not work properly. 
Can you elaborate?

Author: @pritidesai 
> What does properly allude to here?

PiplineRun controller has hard code dependencies on status report of CustomRun controller: See [the code](https://github.com/tektoncd/pipeline/blob/38c739a5d0d9faff8c362462f8b6beee0458d978/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L258-L260), which [is used to decide if the CustomRun is failed or not](https://github.com/tektoncd/pipeline/blob/38c739a5d0d9faff8c362462f8b6beee0458d978/pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go#L148).

@ScrapCodes 
> To ensure that cancellation is required, are we planning to have validations now or in future? if no, is this just a documentation change?

Yes I think we'll have code validations in the future, I opened an issue for it https://github.com/tektoncd/pipeline/issues/5700

> do we support pipeline level Timeout for custom task? I remember it was removed?

There are two kinds of timeout for Custom Task:
- One is [`PipelineTask.Timeout`](https://github.com/tektoncd/pipeline/blob/b15a80e5ab9299386024578ac25b1f42c4d364ac/pkg/apis/pipeline/v1beta1/pipeline_types.go#L229), which is passed to `CustomRunSpec.Timeout`.
- The other is [`PipelineRunSpec.Timeouts`](https://github.com/tektoncd/pipeline/blob/b15a80e5ab9299386024578ac25b1f42c4d364ac/pkg/apis/pipeline/v1beta1/pipelinerun_types.go#L247). 

What we removed is the first one. i.e. we stopped PipelineRun controller to modify CustomRun.status.Conditions upon CustomRun timeout -- (CustomRun executes more than `CustomRunSpec.Timeout`, see https://github.com/tektoncd/pipeline/pull/5658).

> Pipeline Controller sets the `spec.Status` and `spec.StatusMessage` to signal `CustomRuns` about the `Cancellation`, while `CustomRun` controller **MUST** update its `status.conditions` as following once noticed the change on `spec.Status`.

Reviewer: When does pipeline controller set the status about the cancellation? Is it only when it times out or does it also include all types of cancellation that a pipelineRun supports?

https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#cancelling-a-pipelinerun
https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#gracefully-cancelling-a-pipelinerun
https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#gracefully-stopping-a-pipelinerun

Author: Thanks for asking Priti! It includes `Timeout` and `Cancellation` ([cancelling-a-pipelinerun](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#cancelling-a-pipelinerun) and [gracefully-cancelling-a-pipelinerun](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#gracefully-cancelling-a-pipelinerun)[^1], notice [gracefully-stopping-a-pipelinerun](https://github.com/tektoncd/pipeline/blob/main/docs/pipelineruns.md#gracefully-stopping-a-pipelinerun) is not included.)
 
 [^1]: [`cancelPipelineTaskRuns`](https://github.com/tektoncd/pipeline/blob/b15a80e5ab9299386024578ac25b1f42c4d364ac/pkg/reconciler/pipelinerun/cancel.go#L126) is only used in the two kinds of cancellation: [Cancelling a PipelineRun](https://github.com/tektoncd/pipeline/blob/b15a80e5ab9299386024578ac25b1f42c4d364ac/pkg/reconciler/pipelinerun/cancel.go#L9), [Gracefully cancelling a PipelineRun](https://github.com/tektoncd/pipeline/blob/b15a80e5ab9299386024578ac25b1f42c4d364ac/pkg/reconciler/pipelinerun/cancel.go#L205)

Author: Updated the doc. LMK if there's anyother concern!

Reviewer: Thank you @XinruZhang 👍

| Pipeline Signal| `spec.Status` | `spec.StatusMessage` | `status.Conditions` | 
|:--- |:--- |:--- |:---|
|Parent `PipelineRun` is cancelled | `RunCancelled` | `CustomRun cancelled as the PipelineRun it belongs to has been cancelled. `| `Type: Succeeded`<br>`Status:False`<br>`Reason:CustomRunCancelled` |
|`Pipeline` timeouts | `RunCancelled` | `CustomRun cancelled as the PipelineRun it belongs to has timed out.`| `Type: Succeeded`<br>`Status:False`<br>`Reason:CustomRunCancelled`|

> Note: Custom Task implementors **MUST NOT** update the value of `spec.StatusMessage` and `spec.Status`, because Pipeline Controller depends the two fields.

Reviewer: NIT: depends on the two fields - missing on

Author: Thanks! Done~
